orchagent/vnetorch.cpp

```diff
@@ -848,6 +848,162 @@ bool VNetRouteOrch::removeNextHopGroup(const string& vnet, const NextHopGroupKey
+                                       NextHopGroupKey& nexthops,
+                                       IpPrefix& ipPrefix)
+{
+    NextHopGroupKey nhg_custom("", true);
+    set<NextHopKey> next_hop_set = nexthops.getNextHops();
+    for (auto it : next_hop_set)
+    {
+        if(monitor_info_.find(vnet) != monitor_info_.end() &&
+            monitor_info_[vnet].find(ipPrefix) != monitor_info_[vnet].end())
+        {
+            for (auto monitor : monitor_info_[vnet][ipPrefix])
+            {
+                if (monitor.second.endpoint == it)
+                {
+                    if (monitor.second.state == MONITOR_SESSION_STATE_UP)
+                    {
+                        // monitor session exists and is up
+                        nhg_custom.add(it);
+
+                    }
+                    continue;
+                }
+            }
+        }
+    }
+    return nhg_custom;
+}
+
+bool VNetRouteOrch::selectNextHopGroup(const string& vnet,
+                                       NextHopGroupKey& nexthops_primary,
+                                       NextHopGroupKey& nexthops_secondary,
+                                       const string& monitoring,
+                                       IpPrefix& ipPrefix,
+                                       VNetVrfObject *vrf_obj,
+                                       NextHopGroupKey& nexthops_selected,
+                                       const map<NextHopKey, IpAddress>& monitors)
+{
+    if(nexthops_secondary.getSize() != 0 && monitoring == "custom")
+    {
+        auto it_route =  syncd_tunnel_routes_[vnet].find(ipPrefix);
+        if (it_route == syncd_tunnel_routes_[vnet].end())
+        {
+            setEndpointMonitor(vnet, monitors, nexthops_primary, monitoring, ipPrefix);
+            setEndpointMonitor(vnet, monitors, nexthops_secondary, monitoring, ipPrefix);
+        }
+        else
+        {
+            if (it_route->second.primary != nexthops_primary)
+            {
+                setEndpointMonitor(vnet, monitors, nexthops_primary, monitoring, ipPrefix);
+            }
+            if (it_route->second.secondary != nexthops_secondary)
+            {
+                setEndpointMonitor(vnet, monitors, nexthops_secondary, monitoring, ipPrefix);
+            }
+            nexthops_selected = it_route->second.nhg_key;
+            return true;
+        }
+
+        NextHopGroupKey nhg_custom = createActiveNHSet( vnet, nexthops_primary, ipPrefix);
+        if (!hasNextHopGroup(vnet, nhg_custom))
+        {
+            if (!createNextHopGroup(vnet, nhg_custom, vrf_obj, monitoring))
+            {
+                SWSS_LOG_WARN("Failed to create Primary based custom next hop group. Cannot proceed.");
+                delEndpointMonitor(vnet, nexthops_primary, ipPrefix);
+                delEndpointMonitor(vnet, nexthops_secondary, ipPrefix);
+                monitor_info_[vnet].erase(ipPrefix);
```

Comment:
Why can't we move this inside delEndpointMonitor?

Reply:
because of 2 reasons.
1) both primary and secondary are associated with the same prefix. We can go around it by checking if the contents are empty.
2) when we are dealing with the case of route update with endpoint swap from primary to secondary, we need to defer this deletion as we need to reuse the same endpoints. and putting the erase in the delEndpointMonitor causes problems as we are not aware of route update in that context.